File: solutions/csharp/rest-api/1/RestApi.cs
using System;
using System.Collections.Immutable;
using System.Text.Json;
using System.Text.Json.Serialization;

using Debts = System.Collections.Immutable.ImmutableSortedDictionary<string, decimal>;
using UserPredicate = System.Func<User, bool>;

public class RestApi
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
    };

    private readonly List<User> _users;
    private readonly Dictionary<string, int> _userIndex;

    public RestApi(string database)
    {
        _users = Deserialize<List<User>>(database);
        _userIndex = UserIndex(_users);
    }

    public string Get(string url, string? payload = null)
    {
        object response = (url, payload) switch
        {
            ("/users", null) => GetUsers(null),
            ("/users", string usersPayload) => GetUsers(Deserialize<GetUsersRequest>(usersPayload)),
            _ => new { }
        };

        return Serialize(response);
    }

    public string Post(string url, string payload)
    {
        object response = url switch
        {
            "/add" => AddUser(Deserialize<AddUserRequest>(payload)),
            "/iou" => RegisterLoan(Deserialize<IouRequest>(payload)),
            _ => new { }
        };

        return Serialize(response);
    }

    private IEnumerable<User> GetUsers(GetUsersRequest? request)
    {
        UserPredicate filter = request switch
        {
            null => _ => true,
            _ => user => request.Users.Contains(user.Name)
        };

        return _users.Where(filter);
    }

    private User AddUser(AddUserRequest request)
    {
        var user = new User(request.User, Debts.Empty, Debts.Empty, 0);

        _users.Add(user);
        _userIndex[user.Name] = _users.Count - 1;

        return user;
    }

    private IEnumerable<User> RegisterLoan(IouRequest request)
    {
        var lender = UserNamed(request.Lender);
        var borrower = UserNamed(request.Borrower);

        var repayment = Math.Min(request.Amount, lender.AmountOwedTo(borrower));
        var newLoan = request.Amount - repayment;

        var newLender = lender.Repay(borrower, repayment).LendTo(borrower, newLoan);
        var newBorrower = borrower.ReceiveRepaymentFrom(lender, repayment).BorrowFrom(lender, newLoan);

        Replace(newLender);
        Replace(newBorrower);

        return new[] { newLender, newBorrower }.OrderBy(user => user.Name);
    }

    private User UserNamed(string name) => _users[_userIndex[name]];

    private void Replace(User updated)
    {
        _users[_userIndex[updated.Name]] = updated;
    }

    private static T Deserialize<T>(string json) => JsonSerializer.Deserialize<T>(json, JsonOptions)!;

    private static string Serialize(object value) => JsonSerializer.Serialize(value, JsonOptions);

    private static Dictionary<string, int> UserIndex(IEnumerable<User> users) =>
        users
        .Select((user, index) => (user.Name, index))
        .ToDictionary(item => item.Name, item => item.index);
}

record User(
    string Name,
    Debts Owes,
    [property: JsonPropertyName("owed_by")]
    Debts OwedBy,
    decimal Balance
)
{
    public decimal AmountOwedTo(User user) => Owes.GetValueOrDefault(user.Name);

    public User Repay(User user, decimal amount) => WithOwes(user, -amount);

    public User ReceiveRepaymentFrom(User user, decimal amount) => WithOwedBy(user, -amount);

    public User LendTo(User user, decimal amount) => WithOwedBy(user, amount);

    public User BorrowFrom(User user, decimal amount) => WithOwes(user, amount);

    private User WithOwes(User user, decimal amount) => WithDebts(UpdateDebt(Owes, user, amount), OwedBy);

    private User WithOwedBy(User user, decimal amount) => WithDebts(Owes, UpdateDebt(OwedBy, user, amount));

    private User WithDebts(Debts owes, Debts owedBy) =>
        this with
        {
            Owes = owes,
            OwedBy = owedBy,
            Balance = CalculateBalance(owes, owedBy)
        };

    private static Debts UpdateDebt(Debts debts, User user, decimal amount)
    {
        var newAmount = debts.GetValueOrDefault(user.Name) + amount;

        return newAmount == 0 ? debts.Remove(user.Name) : debts.SetItem(user.Name, newAmount);
    }

    private static decimal CalculateBalance(Debts owes, Debts owedBy) => owedBy.Values.Sum() - owes.Values.Sum();

}

record AddUserRequest(string User);

record GetUsersRequest(string[] Users);

record IouRequest(string Lender, string Borrower, decimal Amount);

File: solutions/python/rest-api/1/rest_api.py
"""A small REST-like API for tracking IOUs."""

import json


class RestAPI:
    """Track users and debts between them."""

    def __init__(self, database=None):
        self.users = {
            user["name"]: user
            for user in (database or {"users": []})["users"]
        }

    def get(self, url, payload=None):
        """Handle a GET request."""
        if url == "/users":
            requested_users = self.users

            if payload is not None:
                names = json.loads(payload)["users"]
                requested_users = { name: self.users[name] for name in names }

            return json.dumps({"users": self._sorted_users(requested_users)})

        return None

    def post(self, url, payload=None):
        """Handle a POST request."""
        data = json.loads(payload)

        if url == "/add":
            user = self._new_user(data["user"])
            self.users[user["name"]] = user
            return json.dumps(user)

        if url == "/iou":
            self._lend(data["lender"], data["borrower"], data["amount"])
            updated_users = { name: self.users[name] for name in (data["lender"], data["borrower"]) }
            return json.dumps({"users": self._sorted_users(updated_users)})

        return None

    @staticmethod
    def _new_user(name):
        return {"name": name, "owes": {}, "owed_by": {}, "balance": 0.0}

    @staticmethod
    def _sorted_users(users):
        return [
            users[name]
            for name in sorted(users)
        ]

    def _lend(self, lender_name, borrower_name, amount):
        lender = self.users[lender_name]
        borrower = self.users[borrower_name]

        amount_owed_to_borrower = lender["owes"].get(borrower_name, 0.0)
        offset = min(amount, amount_owed_to_borrower)
        remaining_amount = amount - offset

        if offset:
            self._decrease_debt(lender, borrower, offset)

        if remaining_amount:
            self._increase_debt(borrower, lender, remaining_amount)

    @staticmethod
    def _decrease_debt(borrower, lender, amount):
        RestAPI._adjust_debt(borrower["owes"], lender["name"], -amount)
        RestAPI._adjust_debt(lender["owed_by"], borrower["name"], -amount)
        borrower["balance"] += amount
        lender["balance"] -= amount

    @staticmethod
    def _increase_debt(borrower, lender, amount):
        RestAPI._adjust_debt(borrower["owes"], lender["name"], amount)
        RestAPI._adjust_debt(lender["owed_by"], borrower["name"], amount)
        borrower["balance"] -= amount
        lender["balance"] += amount

    @staticmethod
    def _adjust_debt(debts, name, amount):
        debts[name] = debts.get(name, 0.0) + amount

        if debts[name] == 0:
            debts.pop(name)
